In [2]:
import math
import numpy as np

https://converged.yt/primel/ 

In [1]:
max_num = 99999
 
def get_primes():
    primes = []
    for num in range(3,max_num,2):
        if all(num%i!=0 for i in range(3,int(math.sqrt(num))+1, 2)):
            primes.append(num)
    return primes

In [3]:
primes = get_primes()

filtered = list(filter(lambda primes: primes >= 9999, primes)) # filter primes with <5 digits

primes = filtered

In [4]:
## best primes for initial guess (5 different digits)
start_primes = []
for prime in primes:
    digit_list = [int(x) for x in str(prime)] # creat list of int
    #print(len(list(dict.fromkeys(digit_list))))
    if len(list(dict.fromkeys(digit_list))) == 5: # only for 5 different digits append
        start_primes.append(prime)
print(start_primes[-5:])  

        

[98621, 98627, 98641, 98713, 98731]


In [22]:
digit_count = np.zeros(10)

for prime in primes:
    digit_list = [int(x) for x in str(prime)] # creat list of int
    for digit in digit_list:
        digit_count[digit] = digit_count[digit] + 1
for i, digit in enumerate(digit_count):       
    print(i, ": ", digit_count[i] / len(primes))
    
# best digits to include for start guess: 1, 3, 7, 9

incl_numbers = [1, 3, 7, 9, 2]

temp_primes = primes
for incl_number in incl_numbers:
    best_start_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                best_start_primes.append(prime)
    temp_primes = best_start_primes

best_start_primes = temp_primes
best_start_primes = list(dict.fromkeys(best_start_primes)) 

print(best_start_primes)

0 :  0.298098768384551
1 :  0.6782255171589142
2 :  0.4203037187612101
3 :  0.6638765993064689
4 :  0.40798756427119454
5 :  0.4132488341504245
6 :  0.4032045916537128
7 :  0.6600502212124836
8 :  0.39925863924429034
9 :  0.6557455458567499
[12379, 12739, 12973, 13297, 13729, 17239, 17293, 17923, 19237, 19273, 21379, 21397, 21739, 21937, 23197, 23719, 23917, 23971, 29137, 29173, 31729, 32719, 32917, 32971, 39217, 71293, 71329, 72139, 72931, 73291, 79231, 91237, 92173, 92317, 97213, 97231]


In [24]:
# TRY:       97231  
# YELLOW:    2, 3, 1
# GREY:      7


results = [9, np.NaN, np.NaN, np.NaN, np.NaN] # insert np.NaN for not correct digits, and correct digits in the list
incl_numbers = [2, 3, 1] # put yellow marked digits, and digits you havent tried yet
excl_numbers = [7] # put grey marked digits in here

temp_primes = []

for i, result in enumerate(results):
    if result is not np.NaN:
        if not temp_primes:
            for prime in primes:
                if int(str(prime)[i]) is result:
                    temp_primes.append(prime)
        else:
            new_primes = []
            for prime in temp_primes:
                if int(str(prime)[i]) is result:
                    new_primes.append(prime)
            temp_primes = new_primes

if all(result is np.NaN for result in results): temp_primes = primes            

for incl_number in incl_numbers:
    new_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                new_primes.append(prime)
        temp_primes = new_primes
        
for excl_number in excl_numbers:
    new_primes = []
    for prime in temp_primes:
        append = True
        for digit in str(prime):
            if int(digit) == excl_number:
                append = False 
        if append:    
            new_primes.append(prime)
        temp_primes = new_primes
        
temp_primes = list(dict.fromkeys(temp_primes))            
print(temp_primes) # prints possilbe guesses, use one which includes digits you havent tried yet

[91243, 91253, 91283, 91423, 91823, 92143, 92153, 92311, 92381, 92413, 92431, 92831, 93241, 93251, 93281, 94321, 95213, 95231, 98123, 98213, 98321]


In [25]:
# TRY:       95213  
# YELLOW:    2, 1
# GREY:      5


results = [9, np.NaN, np.NaN, np.NaN, 3] # insert np.NaN for not correct digits, and correct digits in the list
incl_numbers = [2, 3, 1] # put yellow marked digits, and digits you havent tried yet
excl_numbers = [7, 5] # put grey marked digits in here

temp_primes = []

for i, result in enumerate(results):
    if result is not np.NaN:
        if not temp_primes:
            for prime in primes:
                if int(str(prime)[i]) is result:
                    temp_primes.append(prime)
        else:
            new_primes = []
            for prime in temp_primes:
                if int(str(prime)[i]) is result:
                    new_primes.append(prime)
            temp_primes = new_primes

if all(result is np.NaN for result in results): temp_primes = primes            

for incl_number in incl_numbers:
    new_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                new_primes.append(prime)
        temp_primes = new_primes
        
for excl_number in excl_numbers:
    new_primes = []
    for prime in temp_primes:
        append = True
        for digit in str(prime):
            if int(digit) == excl_number:
                append = False 
        if append:    
            new_primes.append(prime)
        temp_primes = new_primes
        
temp_primes = list(dict.fromkeys(temp_primes))            
print(temp_primes) # prints possilbe guesses, use one which includes digits you havent tried yet

[91243, 91283, 91423, 91823, 92143, 92413, 98123, 98213]


In [26]:
# TRY:       91283  
# YELLOW:    1, 2, 8
# GREY:      


results = [9, np.NaN, np.NaN, np.NaN, 3] # insert np.NaN for not correct digits, and correct digits in the list
incl_numbers = [2, 3, 1, 8] # put yellow marked digits, and digits you havent tried yet
excl_numbers = [7, 5] # put grey marked digits in here

temp_primes = []

for i, result in enumerate(results):
    if result is not np.NaN:
        if not temp_primes:
            for prime in primes:
                if int(str(prime)[i]) is result:
                    temp_primes.append(prime)
        else:
            new_primes = []
            for prime in temp_primes:
                if int(str(prime)[i]) is result:
                    new_primes.append(prime)
            temp_primes = new_primes

if all(result is np.NaN for result in results): temp_primes = primes            

for incl_number in incl_numbers:
    new_primes = []
    for prime in temp_primes:
        for digit in str(prime):
            if int(digit) == incl_number:
                new_primes.append(prime)
        temp_primes = new_primes
        
for excl_number in excl_numbers:
    new_primes = []
    for prime in temp_primes:
        append = True
        for digit in str(prime):
            if int(digit) == excl_number:
                append = False 
        if append:    
            new_primes.append(prime)
        temp_primes = new_primes
        
temp_primes = list(dict.fromkeys(temp_primes))            
print(temp_primes) # prints possilbe guesses, use one which includes digits you havent tried yet

[91283, 91823, 98123, 98213]


With exclusion principle: 98123

### ToDo
* Get most frequent digits for initial guess --> Done
* Don't use already tried digits --> Done (grey digits)
* Don't repeat already used position for yellow